### DISCUSSION TITLE
0.51: kubernetes scanner requires role permissions

### DISCUSSION THREAD

**Discussion by jkroepke** [aquasecurity/trivy#6653] [Bugs]
### Description

I'm running trivy with an limited scope of permissions

Before 0.51, i'm runing trivy with following settings:

```bash
      trivy --quiet -q kubernetes --report=all \
        --components workload \
        --disable-node-collector \
        --cache-dir /tmp/.trivycache/ \
        --no-progress \
        --ignore-unfixed \
        --exit-code 0 \
        --slow \
        --format table \
        --scanners vuln \
        --ignorefile /.trivyignore \
        --vuln-type os \
        -n opsstack \
        -o "/scans/report.all.html" \
        deploy,sts,ds;
```

with works fine. After upgrade, I'm using the following command

```
      trivy kubernetes --report=all \
        --disable-node-collector \
        --cache-dir /tmp/.trivycache/ \
        --no-progress \
        --ignore-unfixed \
        --exit-code 0 \
        --slow \
        --format table \
        --scanners vuln \
        --ignorefile /.trivyignore \
        --vuln-type os \
        --include-namespaces opsstack \
        -o "/scans/report.all.html" \
        --include-kinds deploy,sts,ds;
```

I'm getting an error that trivy has no access to roles. However, trivy should not ask for roles.

### Desired Behavior

Except that trivy is scanning the workload of deployments, statefulsets and daemonsets

### Actual Behavior

2024-05-07T15:49:56Z    FATAL   Fatal error     get k8s artifacts error: failed listing resources for gvr: rbac.authorization.k8s.io/v1, Resource=roles - roles.rbac.authorization.k8s.io is forbidden: User "system:serviceaccount:opsstack:opsstack-trivy" cannot list resource "roles" in API group "rbac.authorization.k8s.io" at the cluster scope

### Reproduction Steps

```bash
1. Setup trivy using a context with namespace access only
2. Run the command above
3.
...
```

### Target

Kubernetes

### Scanner

Vulnerability

### Output Format

Table

### Mode

Standalone

### Debug Output

```bash
2024-05-07T15:51:34Z    DEBUG   Parsed severities       severities=[UNKNOWN LOW MEDIUM HIGH CRITICAL]
2024-05-07T15:51:34Z    DEBUG   Ignore statuses statuses=[0 1 2 4 5 6 7]
2024-05-07T15:51:40Z    FATAL   Fatal error
  - get k8s artifacts error:
    github.com/aquasecurity/trivy/pkg/k8s/commands.clusterRun
        /home/runner/work/trivy/trivy/pkg/k8s/commands/cluster.go:49
  - failed listing resources for gvr: rbac.authorization.k8s.io/v1, Resource=roles - roles.rbac.authorization.k8s.io is forbidden: User "system:serviceaccount:opsstack:opsstack-trivy" cannot list resource "roles" in API group "rbac.authorization.k8s.io" at the cluster scope
```

### Operating System

Linux

### Version

```bash
Version: 0.51.1
```

### Checklist

- [ ] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by chen-keinan** [comment]
@jkroepke this functionality is not new, trivy list roles/rolebinding and clusterroles/clusterrolebinding in order to assess RBAC resources.
I can suggest two things:
-  use `--exclude-namespaces` or `include-namespaces` flags to target specific RBAC namespaces
- use `--exclude-kinds clusterroles,roles` to exclude these resources from scanning

**Reply by jkroepke** [reply]
@chen-keinan if you mention, that is not new - why it happens since 0.51?

> use --exclude-kinds clusterroles,roles to exclude these resources from scanning

Running the command

```
      trivy kubernetes --report=all \
        --disable-node-collector \
        --cache-dir /tmp/.trivycache/ \
        --no-progress \
        --ignore-unfixed \
        --exit-code 0 \
        --slow \
        --format table \
        --scanners vuln \
        --ignorefile /.trivyignore \
        --vuln-type os \
        --include-namespaces opsstack \
        -o "/scans/report.all.html" \
        --exclude-kinds clusterroles,roles
```

> 2024-05-09T13:05:35Z    FATAL   Fatal error     get k8s artifacts error: failed listing resources for gvr: rbac.authorization.k8s.io/v1, Resource=roles - roles.rbac.authorization.k8s.io is forbidden: User "system:serviceaccount:opsstack:opsstack-trivy" cannot list resource "roles" in API group "rbac.authorization.k8s.io" at the cluster scope: Azure does not have opinion for this user.


> trivy list roles/rolebinding and clusterroles/clusterrolebinding in order to assess RBAC resources.

How to avoid that? How I can restore the `--components workload` behavior from 0.50?

**Reply by chen-keinan** [reply]
@jkroepke use the following flags`--disable-node-collector` and `--exclude-namespaces kube-system` in conjunction also you can add `--exclude-kinds serviceaccount,roles ` if needed

**Reply by jkroepke** [reply]
The flag `--disable-node-collector` is already included.

The flag `--include-namespaces opsstack` should already cover `--exclude-namespaces kube-system`?

Same question for the initial suggestion: I'm original using  `--include-kinds deploy,sts,ds` what should exclude everything, like role, sa, clusterrole and more.

Here is the command which includes your suggestions:

```
 trivy kubernetes --report=all \
        --disable-node-collector \
        --cache-dir /tmp/.trivycache/ \
        --no-progress \
        --ignore-unfixed \
        --exit-code 0 \
        --slow \
        --format table \
        --scanners vuln \
        --ignorefile /.trivyignore \
        --vuln-type os \
        --exclude-namespaces kube-system \
        --exclude-kinds clusterroles,serviceaccount,roles
```

and results into the same error again:

```
2024-05-09T13:49:47Z    FATAL   Fatal error     get k8s artifacts error: failed listing resources for gvr: rbac.authorization.k8s.io/v1, Resource=roles - roles.rbac.authorization.k8s.io is forbidden: User "system:serviceaccount:opsstack:opsstack-trivy" cannot list resource "roles" in API group "rbac.authorization.k8s.io" at the cluster scope: Azure does not have opinion for this user.
```

**Reply by jkroepke** [reply]
@chen-keinan seems like there is an bug in trivy.

**Reply by chen-keinan** [reply]
@jkroepke thanks for the input converting to issue and I'll will create a fix for it.

**Comment by chen-keinan** [comment]
issue opened #6692

**Reply by jkroepke** [reply]
Hi @chen-keinan 

i try trivy 0.52 today which that command:

```
trivy --quiet -q kubernetes --report=all \
        --disable-node-collector \
        --cache-dir /tmp/.trivycache/ \
        --no-progress \
        --ignore-unfixed \
        --exit-code 0 \
        --slow \
        --format table \
        --scanners vuln \
        --ignorefile /.trivyignore \
        --vuln-type os \
        --include-namespaces opsstack \
        -o "/scans/report.all.html" \
        --include-kinds deploy,sts,ds;
```

Which no results into an new error:

> 2024-06-03T16:34:29Z	FATAL	Fatal error	get k8s artifacts error: nodes is forbidden: User "system:serviceaccount:opsstack:opsstack-trivy" cannot list resource "nodes" in API group "" at the cluster scope

Mention that `--disable-node-collector` is set and `node` is not inside `--include-kinds`. There no reason for asking nodes.

Efficiently, I'm just missing the `components workload` flag, where trivy takes care of the workload not care about the infrastructure. In my case, trivy should just report os vuln of running images in a given namespace, not more.

**Reply by Mafo75** [reply]
Hi @chen-keinan ,

we have the same problem, I tested yesterday and with version 0.53, but it seems that the bug is not fixed.

Do you have a plan to fix it or we have to use the old version ?

**Reply by chen-keinan** [reply]
@Ill have a look and update you

**Reply by chen-keinan** [reply]
@Ill have a look and update you

**Reply by chen-keinan** [reply]
@jkroepke the reason we are listing nodes is for looking vulnerabilities inside node components: kubelet, containerd, kube-proxy and etc.
I'm creating a fix on this to make sure it will not fail on permission for it.

**Reply by jkroepke** [reply]
Hey @chen-keinan 


what about an flag that only validate the workload?

Like we had in 0.50 and below

Can we reopen this discussion until the issue is resolved?

**Reply by chen-keinan** [reply]
I have added a fix below which should sort out functionality, no need other workaround

**Reply by chen-keinan** [reply]
lets have discussion on the issue #6692, its open now

**Comment by chen-keinan** [comment]
I have opened the issue #6692 following investigation

Added a fix: #7107
